### PR TITLE
Add metric name validation

### PR DESCRIPTION
**Pull request by pgomulka**
This commit adds a minimum metric name validation which checks:
- metric name starts with `es.` prefix
- metric name is using `.` as a separator of elements
- metric name is using characters from a white list
- validate min number of elements = 3 elements ( prefix, group and the suffix name)
- validate max number of elements and max characters per element
- validate the suffix element in a metric name to be from the enumerated allow list

### CODE REVIEW

**Comment by pgomulka** [comment]
this PR also aligns the already existing metric names to pass the validation. I am ping people who introduced the as I likely will break your dashboards in kibana.
https://github.com/elastic/elasticsearch/pull/103388/files#diff-3b0ab75398e06bcff863305ac1b6e725917476eb32d5c2ccbf09e1ed884405f7

and 
https://github.com/elastic/elasticsearch/pull/103388/files#diff-74956e42dd1188b5474ec53450471777baa197a6bceb325716e31ec00d31c3f8
@ywangd 

https://github.com/elastic/elasticsearch/pull/103388/files#diff-701cea9e04fe2f63a6e0d573100f0dff75e97c8eb299b00ded7662860c5922ef
@mattc58 


https://github.com/elastic/elasticsearch/pull/103388/files#diff-b6b5bf4262dc56bace4fb61258984a3519f448349d9efd687a64ac186418e1e5
but this is already being changed in a separate PR https://github.com/elastic/elasticsearch/pull/103089

**Comment by ywangd** [comment]
Thanks @pgomulka Let me ping more folks for awareness.

@idegtiarenko for [desired balance](https://github.com/elastic/elasticsearch/pull/103388/files#diff-3b0ab75398e06bcff863305ac1b6e725917476eb32d5c2ccbf09e1ed884405f7) metrics name changes
@JVerwolf for [blob cache](https://github.com/elastic/elasticsearch/pull/103388/files#diff-74956e42dd1188b5474ec53450471777baa197a6bceb325716e31ec00d31c3f8) metrics name changes

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java:118]: Ack. I will update dashboards once this is merged

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java
Comment (idegtiarenko) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:18]: Consider this to `ALLOWED_SUFFIXES` or `PERMITTED_SUFFIXES`

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java
Comment (idegtiarenko) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:33]: Is there overall metric name length limit? (not one per element)

**Review by idegtiarenko — APPROVED** [review]

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java
Comment (pgomulka) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:33]: yes, it is 255characters. It is being verified by an agent itself

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java
Comment (ldematte) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:33]: Can you add javadoc, and a link to `modules/apm/NAMING.md`?

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java
Comment (ldematte) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:34]: I think it would be better to have it here too, so we can fail here consistently

File: server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java:209]: Trying to understand "status" here: why not "count" (0/1)?

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:119]: Should be singular (operation.count)

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:137]: Same here (singular)

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:155]: Nit: isn't "age" redundant? earliest_last_modified.time looks better, IMO

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:227]: Are we sure this is total and not count? I see the description names "Current". Maybe is it the description that needs update instead?

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:254]: Same

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:317]: Same

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:362]: Same as above, it is probably just `size`?

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:371]: Same: probably just `count`?

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:389]: Same: just `limit` (and it needs to be added to the suffixes too probably)

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java
Comment (ldematte) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:23]: Is this `total` instead?

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java
Comment (ldematte) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:28]: Same (possibly `total`?)

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java
Comment (ldematte) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:35]: Wrong unit of measure?

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (stu-elastic) [inline server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:121]: Should this change to "count"?

**Review by stu-elastic — APPROVED** [review]
lgtm once my and @ldematte's comments are addressed.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java
Comment (rjernst) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:33]: Could this be static? The validation methods seem to be stateless, no need to have an instance?

File: server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java:209]: this is a workaround for a lack of enum/boolean support in OTEL. 
count would indicate that this value will be increasing.

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:119]: hmm are you sure? it actually feels ok to me

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:155]: fair point, I suspect this was meant to replicate already existing metrics returned from that node-stats api
I will rename.

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:227]: good catch. I think this has to keep the current subgroup to indicate this is about 'in-flight indexing operations'

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:254]: I am not sure, the javadoc for   `getIndexFailedCount`  says `The number of failed indexing operations`
both `total` and `count` feel ok to me..
maybe it should not be a longGauge but asyncCounter

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:317]: I think this can be just .size

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:362]: agree

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:389]: how about memory_limit.total ?

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java
Comment (pgomulka) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:23]: count and total feel very similar to me. I would use `total` for async counter or gauge (reporting a cumulative value)
and count for `deltas` like here.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java
Comment (pgomulka) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:28]: count and total feel very similar to me. I would use total for async counter or gauge (reporting a cumulative value)
and count for deltas like here.

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:121]: translong.operations.count feels addivite to me. perhaps it should be an async counter instead

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java
Comment (pgomulka) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:33]: I tend to start with a regular class instead of a static util. Static utilities are creating tight coupling and prevent testing in isolation.
However I have moved the use of `MetricNameValidator` to `AbstractInstrument` and refactored all the test metric names to pass the rules so there probably will never be need to test instruments in isolation of the metric name validation.

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java]: Yes, from both our guidelines and https://opentelemetry.io/docs/specs/semconv/general/metrics/.
It could be either `operations` or `operation.count`, but since we decided for fixed suffixes, it should be the second.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java
Comment (ldematte) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:23]: Count should for the current number of a fluctuating quantity, e.g. the number of tabs _currently_ opened in your browser is `browser.window.open.count`.
Total should be for an always increasing quantity, e.g. the number of times you have been oncall `oncall.total`

This looks like a total to me (we look at all the cache misses and we keep increasing the counter).

Comment (pgomulka) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:23]: as we discussed offline.
I will change `count` to be `current`. The intention is to use this for values of current snapshot/sample. That is for non-monotonic metrics (number of tabs can go up and down) Does not matter if it is additive or not. 
`total` will be always monotonic metrics (a counter). I will remove the word of `sum` as it might be confusing if the value reported should cumulative or delta.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java
Comment (pgomulka) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:28]: changed to total

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java]: I have changed this to total.
unit here is misleading. we often use just `count` as counters are often unit-less.
I wonder if we should not  allow empty strings here

**Comment by pgomulka** [comment]
note for reviewers( @rjernst  @stu-elastic ). Please take a look at `total` vs `current` vs `count`
I have changed the naming guide to be more clear about this
```
* Use `total` as a suffix for monotonic metrics (always increasing counter) (e.g. <code>es.indices.docs.deleted.<strong>total</strong></code>)
  * Note: even though async counter is reporting a total cumulative value, it is till monotonic.
* Use `current` to represent the non-monotonic metrics (like gauges, upDownCounters) (e.g. if we have `es.process.jvm.classes.loaded`, we will express the number of classes currently loaded by the JVM as <code>es.process.jvm.classes.loaded.<strong>count</strong></code>, and the total number of classes loaded since the JVM started as <code>es.process.jvm.classes.loaded.<strong>total</strong></code>
```
I felt like `count` was often misleading for metric names used in gauges. Hence a rename to `current`
but `current` also is not ideal see https://github.com/elastic/elasticsearch/pull/102371/files/01260a3599eef59426d505b47d4193d0e27b3339#diff-2835e151b1e7435ff4760a9263e0ecbc359597a9ff02005fc7fec099d38549e6
for instance:
 `es.thread_pool." + name + ".threads.current` (gauge)  
or
` es.thread_pool." + name + ".threads.active.current` (gauge)
those does not indicate that they are a `unit-ness` numbers.

**the alternative we discussed with @ldematte  (pretty much reverse the count and total)**
```
* Use `count` as a suffix for monotonic metrics (always increasing counter) (e.g. <code>es.indices.docs.deleted.<strong>total</strong></code>)
  * Note: even though async counter is reporting a total cumulative value, it is till monotonic.
* Use `total` to represent the non-monotonic metrics (like gauges, upDownCounters). The overall cumulative value of a measured resource at the sample time
```

File: server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java]: Yes, size is best here.

**Comment by stu-elastic** [comment]
Re: `count` vs `total`.  I know that the desire is to be able to differeniate between sample metrics like gauges and additive metrics like counters.

However, those two look very similar and we are adding cognitive load on authors instrumenting their code by making them pick between the two.  While differentiating between metric types is a useful goal, I do not think it's worth having two very similar units.

We are not enforcing per-instrument-type naming guidelines so we should expect regular confusion over which to pick.

I suggest dropping `total` and going with `count` and also removing `current`, as it is not a unit.

We can always add new units but it will be a backwards compatibility issue if we let one unit in and then try to remove it.

**Comment by JVerwolf** [comment]
Question: Why do we need to manually add suffixes to metrics when this could be automated by the metric infrastructure itself? On this note, if we are going to enforce suffixes based on metric types, why not include the metric type as a suffix and remove this extra layer?  e.g. a counter could add `.counter`, a histogram `.histogram`, a gauge adds `.gauge`, etc.

**Comment by rjernst** [comment]
> Why do we need to manually add suffixes to metrics

I think the reasoning is similar to why we don't automatically add the `es.` prefix. We want the metric name to be searchable in the codebase. So if you are looking at a dashboard, and you see the source data is `es.foo.bar.counter`, you should be able to search for that string. We have prior examples eg in xpack settings where the actual setting name is built up and therefore obfuscated, and it makes navigating to where a setting is defined difficult. We don't want to make that same mistake with metrics.

**Comment by rjernst** [comment]
> I suggest dropping total and going with count and also removing current, as it is not a unit.

I think there is a benefit of having different suffixes for cumulative vs current values. These often appear side-by-side. For example, the number of docs indexed. We have the total of all time for the node, and the current number being indexed.

**Comment by ldematte** [comment]
> I think there is a benefit of having different suffixes for cumulative vs current values. These often appear side-by-side. For example, the number of docs indexed. We have the total of all time for the node, and the current number being indexed.

++
While compiling the guidelines, that was exactly the scenario we depicted. You may want to have both (e.g. `es.something.cached.count` + `es.something.cached.total`

`current` was proposed as a more explicit/less ambiguous replacement for `count`

> Why do we need to manually add suffixes to metrics when this could be automated by the metric infrastructure itself?

Another reason: avoid name clashing and decouple the instrument type from the metric name.

That said, I'm happy to go either way: we can drop the difference and just keep `count` for every `unit` measure (and work around the previous case with more elaborate names, e.g. `es.something.cached.current.count` + `es.something.cached.total.count`, only when we need it).
If we go that way, I'd update the guidelines (probably in the same commit, since they now live in the repo).

**Comment by stu-elastic** [comment]
I'm fine with having a separate `count` vs `total` to differentiate the instrument so long as it's easy for metric authors to know when to use which suffix.

**Pull request merged**
